src/redux-middleware/remoteSync.ts

```diff
@@ -0,0 +1,121 @@
+import _ from 'lodash'
+import { ThunkMiddleware } from 'redux-thunk'
+import { decodeContextUrl, expandThoughts } from '../selectors'
+import { equalArrays, hashContext, hashThought, head, keyValueBy, pathToContext } from '../util'
+import { State } from '../util/initialState'
+import { Index, Context, Ref, Parent, Lexeme } from '../types'
+import { reconcile } from '../action-creators'
+import getFirebaseProvider from '../data-providers/firebase'
+
+/** Generates a map of all visible contexts, including the cursor, all its ancestors, and the expanded contexts. */
+const getVisibleContexts = (state: State, expandedContexts: Index<Context>): Index<Context> => {
+
+  const { cursor } = state
+
+  // if there is no cursor, decode the url so the cursor can be loaded
+  // after loading the ranks will be inferred to update the cursor
+  const contextUrl = decodeContextUrl(state, window.location.pathname)
+  const contextCursor = cursor ? pathToContext(cursor) : contextUrl
+
+  return {
+    ...expandedContexts,
+    // generate the cursor and all its ancestors
+    // i.e. ['a', b', 'c'], ['a', 'b'], ['a']
+    ...keyValueBy(contextCursor, (value, i) => {
+      const subcontext = contextCursor.slice(0, contextCursor.length - i)
+      return subcontext.length > 0 ? { [hashContext(subcontext)]: subcontext } : null
+    }),
+  }
+}
+
+/**
+ * Middleware to auto-sync thoughts from firebase. 2 major functions that this serves
+ * 1) Maintains the list(thoughtListeners) of listeners for visible contexts, while deactivating the ones that are hidden.
+ * 2) Builds the thoughtIndex by fetching all lexemes and triggers reconcile with thought and context index updates.
+ */
+const remoteSyncMiddleware: ThunkMiddleware<State> = ({ getState, dispatch }) => {
```

Comment:
Let's use the term `remoteSubscribe` rather than `remoteSync` since it handles subscribing to remote changes and sync could also refer to pull or push.